kafka.go

```diff
@@ -194,84 +189,141 @@ func (kafka *kafkaReader) Seek(ctx context.Context, offset int64) (int64, error)
 // Get the real offset from Kafka. If a non-negative offset is provided this method won't
 // do anything and will return the same offset. If the offset provided is -1/-2 then it will
 // make a call to Kafka to fetch the real offset.
-func (kafka *kafkaReader) getOffset(broker *sarama.Broker, offset int64) (int64, error) {
+func (kafka *kafkaReader) getOffset(ctx context.Context, offset int64) (int64, error) {
 	// An offset of -1/-2 means the partition has no offset state associated with it yet.
 	//  -1 = Newest Offset
 	//  -2 = Oldest Offset
-	if offset == -1 || offset == -2 {
+	if offset != -1 && offset != -2 {
+		return offset, nil
+	}
+
+	errs := make(chan error)
+	val := make(chan int64)
+
+	broker, err := kafka.getLeader(ctx)
+	if err != nil {
+		return 0, err
+	}
+
+	go func() {
 		request := &sarama.OffsetRequest{Version: 1}
 		request.AddBlock(kafka.topic, kafka.partition, int64(offset), 1)
 
 		offsetResponse, err := broker.GetAvailableOffsets(request)
 		if err != nil {
-			return offset, errors.Wrap(err, "failed to fetch available offsets")
+			errs <- errors.Wrap(err, "failed to fetch available offsets")
+			return
 		}
 
 		block := offsetResponse.GetBlock(kafka.topic, kafka.partition)
 		if block == nil {
-			return offset, errors.Wrap(err, "fetching available offsets returned 0 blocks")
+			errs <- errors.Wrap(err, "fetching available offsets returned 0 blocks")
+			return
 		}
 
 		if block.Err != sarama.ErrNoError {
-			return offset, errors.Wrap(err, "fetching available offsets failed")
+			errs <- errors.Wrap(err, "fetching available offsets failed")
+			return
 		}
 
 		if block.Offset != 0 {
-			return block.Offset - 1, nil
+			val <- block.Offset - 1
+		} else {
+			val <- block.Offset
```

Comment:
It seems like all the sends on channels in this goroutine would block if the context is canceled, causing the goroutine to leak.